File: Palace/Book/UI/TPPBookButtonsState.m
//
//  TPPBookButtonsState.m
//  The Palace Project
//
//  Created by Ettore Pasquini on 3/18/21.
//  Copyright © 2021 NYPL Labs. All rights reserved.
//

#import "TPPBookButtonsState.h"
#import "TPPOPDSAcquisitionAvailability.h"
#import "Palace-Swift.h"

TPPBookButtonsState
TPPBookButtonsViewStateWithAvailability(id<TPPOPDSAcquisitionAvailability> const availability)
{
  __block TPPBookButtonsState state = TPPBookButtonsStateUnsupported;

  if (!availability) {
    [TPPErrorLogger logErrorWithCode:TPPErrorCodeNoURL
                              summary:@"Unable to determine BookButtonsViewState because no Availability was provided"
                             metadata:nil];
  }

  [availability
   matchUnavailable:^(__unused TPPOPDSAcquisitionAvailabilityUnavailable *_Nonnull unavailable) {
    state = TPPBookButtonsStateCanHold;
  }
   limited:^(__unused TPPOPDSAcquisitionAvailabilityLimited *_Nonnull limited) {
    state = TPPBookButtonsStateCanBorrow;
  }
   unlimited:^(__unused TPPOPDSAcquisitionAvailabilityUnlimited *_Nonnull unlimited) {
    state = TPPBookButtonsStateCanBorrow;
  }
   reserved:^(__unused TPPOPDSAcquisitionAvailabilityReserved *_Nonnull reserved) {
    state = TPPBookButtonsStateHolding;
  }
   ready:^(__unused TPPOPDSAcquisitionAvailabilityReady *_Nonnull ready) {
    state = TPPBookButtonsStateHoldingFOQ;
  }];

  return state;
}

File: Palace/Book/UI/TPPBookButtonsState.swift
import Foundation

@objc enum TPPBookButtonsState: Int {
  case canBorrow
  case canHold
  case holding
  case holdingFOQ // Front Of Queue: a book that was Reserved and now it's Ready for borrow
  case downloadNeeded
  case downloadSuccessful
  case used
  case downloadInProgress
  case downloadFailed
  case unsupported
}

/// Returns a Borrow, Keep, Hold, Holding, or HoldingFOQ state.
func TPPBookButtonsViewStateWithAvailability(_ availability: TPPOPDSAcquisitionAvailability?) -> TPPBookButtonsState {
  guard let availability = availability else {
    TPPErrorLogger.logError(
      withCode: .noURL,
      summary: "Unable to determine BookButtonsViewState because no Availability was provided",
      metadata: nil
    )
    return .unsupported
  }

  var state: TPPBookButtonsState = .unsupported

  availability.match(
    unavailable: { _ in state = .canHold },
    limited: { _ in state = .canBorrow },
    unlimited: { _ in state = .canBorrow },
    reserved: { _ in state = .holding },
    ready: { _ in state = .holdingFOQ }
  )

  return state
}
